### PR TITLE
Seperate allocs for KAS_READ_ALL

### LINKED ISSUE #175: Change READ_ALL semantics to sequentially reading to separate buffers
In the READ_ALL case we're currently reading the while data block of the file into a single buffer, and then passing out pointers into that buffer when we call read. This adds complexity to the code because we need to reason about what needs freeing etc during close.

Change the behaviour to sequentially reading each array into it's own malloced buffer that we store in the ``item`` struct. These are automatically freed then at close. The most important thing is that we get the size of the fread() calls right. This must be the length of the current array + padding, so basically ``next_item_offset - current_offset``.

This should have identical behaviour then to the current implementation and no changes to the test suite would be required.

We can probably remove the ``self->read_buffer`` attribute.

### PR DESCRIPTION
**Pull request by benjeffery**
Fixes #175

### CODE REVIEW

**Comment by jeromekelleher** [comment]
Odd, not sure why the tests are failing - this looks like it should work. I can take a closer look if that'd be helpful?

**Comment by benjeffery** [comment]
> Odd, not sure why the tests are failing - this looks like it should work. I can take a closer look if that'd be helpful?

Yeah, the C tests are fine! I can only think I'm doing something like reading past the end or something that python is picking up on?

File: c/kastore.c [hunk 1]
```diff
@@ -401,10 +398,24 @@ kastore_read_file(kastore_t *self)
...
         self->items[j].key = self->read_buffer + self->items[j].key_start - offset;
         if (read_all) {
-            self->items[j].array
-                = self->read_buffer + self->items[j].array_start - offset;
+            if (j == self->num_items - 1) {
+                size = self->file_size - self->items[j].array_start;
+            } else {
+                size = self->items[j + 1].array_start - self->items[j].array_start;
+            }
+            self->items[j].array = (char *) malloc(size);
```
Comment (jeromekelleher) [inline c/kastore.c:409]: watch out for malloc 0 here, do ``malloc(size == 0? 1 : size)``.

**Comment by jeromekelleher** [comment]
Hah - looks like ``fread()`` doesn't like being called with a size of 0. Not sure why this only shows up here and not in the non-read-all case, but :shrug:

**Comment by benjeffery** [comment]
> Hah - looks like `fread()` doesn't like being called with a size of 0. Not sure why this only shows up here and not in the non-read-all case, but 🤷

Ah! Nice one.

**Comment by benjeffery** [comment]
@mergifyio rebase

**Comment by benjeffery** [comment]
@jeromekelleher I think this is all good now, had to fix mergify too, damn API churn.

**Review by jeromekelleher — APPROVED** [review]

**Pull request merged**